### PR TITLE
NSPR-1387 : Xrt hbm support

### PR DESCRIPTION
**Pull request by saifuddin-xilinx**
Hi,
Creating this PR for review this HBM changes.
Please review the changes follows:

Userspace changes for XRT layer
User_PF Driver changes for HBM support
New changes support the following functionality:

User can specify a range of HBM banks in the v++ connectivity section, Example:
-sp CU_0.in1:HBM[0:3] . The user needs to specify bank-id information in his host application
as well using Extension Pointer.
If the user doesn't specify anything through the Extension pointer, Then XRT should automatically
detect the HBM banks using the connectivity section present in rtd file.

Further details : https://confluence.xilinx.com/pages/viewpage.action?pageId=205359608

### CODE REVIEW

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c [hunk 1]
```diff
@@ -2519,6 +2519,34 @@ static inline int icap_xmc_free(struct icap *icap)
 	return err == -ENODEV ? 0 : err;
 }
 
+static bool check_group_topo_and_connectivity(struct icap *icap,
+	struct axlf *xclbin)
+{
+    struct connectivity *group_conn = icap->connectivity;
+	struct mem_topology *group_topo = icap->mem_topo;
```
Comment (keryell) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c:2526]: Please avoid this random mix of spaces and tabulation characters. :-(
Use a tool like `clang-format` or configure your editor or IDE.

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_bo.c [hunk 2]
```diff
@@ -484,7 +482,9 @@ int xocl_create_bo_ioctl(struct drm_device *dev,
 		else if (xobj->flags & XOCL_DRM_SHMEM)
 			xobj->pages = drm_gem_get_pages(&xobj->base);
 		else if (xobj->flags & XOCL_CMA_MEM){
-			uint64_t start_addr = drm_p->mm[ddr]->head_node.start + drm_p->mm[ddr]->head_node.size;
+            /* SAIF TODO : Need to rechaek this */
```
Comment (keryell) [inline src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_bo.c:485]: What about configuring your developer environment to use a spell-checker?

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_drm.c [hunk 3]
```diff
@@ -520,14 +520,43 @@ void xocl_mm_update_usage_stat(struct xocl_drm *drm_p, u32 ddr,
 	drm_p->mm_usage_stat[ddr]->bo_count += count;
 }
 
+int xocl_mm_insert_node_range(struct xocl_drm *drm_p, u32 mem_id,
+			      struct drm_mm_node *node, u64 size)
+{
+	uint64_t start_addr = 0;
+	uint64_t end_addr = 0;
+	struct mem_topology    *topology = NULL;
```
Comment (keryell) [inline src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_drm.c:528]: Not clear why so much spaces after `mem_topology`

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c [hunk 4]
```diff
@@ -2519,6 +2519,34 @@ static inline int icap_xmc_free(struct icap *icap)
...
+	const struct axlf_section_header *hdr = get_axlf_section_hdr(icap, xclbin, ASK_GROUP_TOPOLOGY);
+
+	if (!hdr || !group_topo)
+		return false;
+
+    /* If group_topology section presents, overwrite mem_topology with it */
+    memcpy(group_topo, ((const char *)xclbin) + hdr->m_sectionOffset,
```
Comment (maxzhen) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c:2533]: Did you make sure the buffer pointed to by group_topo is big enough for you to copy a new section to it? I'm not sure I understand what you are trying to do here... After the copy, we don't see original mem topo section from the xclbin any more in driver? I thought that we have discussed this before and already made it clear that mem topo is going to be used by xbutil. You can use mem topo section as group topo, if group topo is not found. But you can't use group topo as mem topo, right? Same thing below for connectivity section...

File: src/runtime_src/core/pcie/linux/shim.cpp [hunk 5]
```diff
@@ -1413,6 +1413,44 @@ int shim::xclGetBOProperties(unsigned int boHandle, xclBOProperties *properties)
...
+ */
+int shim::xclGetMemInfo(char **MemInfo)
+{
+    std::string errmsg;
+    size_t size = 0;
+    std::vector<char> membuf;
+    std::vector<char> connbuf;
+
+    /* Get the memory topology Information */
+    mDev->sysfs_get("icap", "mem_topology", errmsg, membuf);
```
Comment (chvamshi-xilinx) [inline src/runtime_src/core/pcie/linux/shim.cpp:1427]: This should be group_topology not mem_topology. right?

File: src/runtime_src/core/pcie/linux/shim.cpp [hunk 6]
```diff
@@ -1413,6 +1413,44 @@ int shim::xclGetBOProperties(unsigned int boHandle, xclBOProperties *properties)
...
+    if (!errmsg.empty()) {
+        xrt_logmsg(XRT_ERROR, "%s: %s", __func__, errmsg);
+        return -EINVAL;
+    }
+
+    /* Get the memory connectivity Information */
+    mDev->sysfs_get("icap", "connectivity", errmsg, connbuf);
```
Comment (chvamshi-xilinx) [inline src/runtime_src/core/pcie/linux/shim.cpp:1434]: Same here. This should be group_connectivity not connectivity. right?

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c (hunk 1 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c]: Will take care of that.

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_drm.c (hunk 3 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_drm.c]: Will fix it.

**Comment by saifuddin-xilinx** [comment]
This branch is out of sync of latest master. Discard this pull request and create another one.

**Pull request closed without merging**